File: algorithm/pigeonhole-sort.js
/**
 * Pigeonhole Sort Implementation in JavaScript
 * 
 * Pigeonhole sort is an extreme version of counting sort. It requires that the
 * range of possible values is small and known. Each value gets its own "pigeonhole".
 * 
 * Time Complexity: O(n + range) where range is the difference between max and min
 * Space Complexity: O(range)
 */

function pigeonholeSort(arr) {
    if (arr.length === 0) return arr;
    
    const min = Math.min(...arr);
    const max = Math.max(...arr);
    const range = max - min + 1;
    
    // Create pigeonholes
    const holes = new Array(range).fill(0);
    
    // Count occurrences
    for (const num of arr) {
        holes[num - min]++;
    }
    
    // Reconstruct sorted array
    const sorted = [];
    for (let i = 0; i < range; i++) {
        while (holes[i] > 0) {
            sorted.push(i + min);
            holes[i]--;
        }
    }
    
    return sorted;
}

// Example usage
if (require.main === module) {
    const arr = [8, 3, 2, 7, 4, 6, 8];
    console.log('Pigeonhole Sort:', pigeonholeSort(arr));
}

module.exports = { pigeonholeSort };

File: algorithm/pigeonhole-sort.py
"""
Pigeonhole Sort Implementation in Python

Pigeonhole sort is an extreme version of counting sort. It requires that the
range of possible values is small and known. Each value gets its own "pigeonhole".

Time Complexity: O(n + range) where range is the difference between max and min
Space Complexity: O(range)
"""


def pigeonhole_sort(arr):
    if len(arr) == 0:
        return arr
    
    min_val = min(arr)
    max_val = max(arr)
    range_val = max_val - min_val + 1
    
    # Create pigeonholes
    holes = [0] * range_val
    
    # Count occurrences
    for num in arr:
        holes[num - min_val] += 1
    
    # Reconstruct sorted array
    sorted_arr = []
    for i in range(range_val):
        while holes[i] > 0:
            sorted_arr.append(i + min_val)
            holes[i] -= 1
    
    return sorted_arr


if __name__ == "__main__":
    arr = [8, 3, 2, 7, 4, 6, 8]
    print("Pigeonhole Sort:", pigeonhole_sort(arr))

